File: Leetcode/linkedlist/234_PalindromeLinkedList/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next==null)
            return true;
        Stack<Integer> res = new Stack<Integer>();
        ListNode slow = head, fast = head;
        while (fast!=null && fast.next!=null){
            res.push(slow.val);
            slow = slow.next;  fast = fast.next.next;
        }
        if(fast!=null)
            slow=slow.next;     //check if it is odd length mid point
        while(slow!=null){
            if (slow.val!=res.pop())
                return false;
            slow = slow.next;
        }
        return true;
    }
}

File: Leetcode/linkedlist/234_PalindromeLinkedList/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        stack<int> stk;
        ListNode* p = head;
        while(p!=nullptr){
            stk.push(p->val);
            p = p->next;
        }
        p = head;
        while(!stk.empty()){
            if(p->val!=stk.top()){
                return false;
            }
            stk.pop();
            p = p->next;
        }
        return true;
    }
};
